docs/integrations/nestjs.md

```diff
@@ -0,0 +1,198 @@
+
+async function bootstrap() {
+  const app = await NestFactory.create(AppModule);
+  // highlight-next-line
+  await DBOS.launch();
+  await app.listen(process.env.PORT ?? 3000);
+}
+bootstrap();
+```
+
+### Register Services With DBOS
+To integrate a Nest.js service with DBOS, your service class must extend the DBOS [ConfiguredInstance](https://docs.dbos.dev/typescript/reference/transactapi/dbos-class#decorating-instance-methods) class. By extending `ConfiguredInstance`, you add your class instance methods to DBOS Transact's internal registry. During [workflow recovery](https://docs.dbos.dev/typescript/tutorials/workflow-tutorial#workflow-versioning-and-recovery), this registry enables DBOS to recover workflows using the right class instance.
+
+Here is an example with a Nest.js service implementing a simple two-step workflow:
+
+```typescript
+// app.service.ts
+import { Injectable } from '@nestjs/common';
+import { PrismaService } from 'nestjs-prisma';
+// highlight-next-line
+import { ConfiguredInstance, DBOS, InitContext } from '@dbos-inc/dbos-sdk';
+
+@Injectable()
+// highlight-next-line
+export class AppService extends ConfiguredInstance {
+  constructor(
+    name: string, // You must provide a name for this class instance to uniquely identify it in DBOS's internal registry.
+    private readonly prisma: PrismaService, // An example service dependency
+  ) {
+    super(name);
+  }
+
+  // Optionally perform some asynchronous setup work
+  async initialize(): Promise<void> {}
+
+  // highlight-next-line
+  @DBOS.workflow()
+  async businessLogic() {
+    await this.step1();
+    await this.step2();
+  }
+
+  // highlight-next-line
+  @DBOS.step()
+  async step1() {
+    ...
+  }
+
+  // highlight-next-line
+  @DBOS.step()
+  async step2() {
+    ...
+  };
+}
+```
+
+### Add Nest.js Providers
+Finally, we need to write the code that Nest will use to instantiate this service during dependency injection. We'll do this with a [custom _Factory Provider_](https://docs.nestjs.com/fundamentals/custom-providers#factory-providers-usefactory). Here is an example:
+
+```typescript
+// app.modules.ts
+import { Module } from '@nestjs/common';
+import { Provider } from '@nestjs/common/interfaces';
+import { AppController } from './app.controller';
+import { AppService } from './app.service';
+import { PrismaService, PrismaModule } from 'nestjs-prisma';
+import { DBOS } from '@dbos-inc/dbos-sdk';
+
+export const dbosProvider: Provider = {
+  provide: AppService,
+  useFactory: (prisma: PrismaService) => {
+    return DBOS.configureInstance(AppService, "dbosService", prisma);
+  },
+  inject: [PrismaService],
+};
+
+@Module({
+  imports: [PrismaModule.forRoot()],
+  controllers: [AppController],
+  providers: [dbosProvider],
+})
+export class AppModule {}
+```
+
+If you need multiple instances of your DBOS class, you must give them distinct names (`dbosService` in this case) and create a dedicated provider for each.
```

Comment:
Why do you need a dedicated provider for each?

Reply:
Because each provider must have a unique token. There is of course "yet another way" to have multiple classes, but I didn't see it in real applications


```typescript
export const dbosProviders: Provider = {
  provide: 'DBOS_SERVICES',
  useFactory: (prisma: PrismaService) => {
    return [
      DBOS.configureInstance(AppService, 'ServiceA', prisma),
      DBOS.configureInstance(AppService, 'ServiceB', prisma),
    ];
  },
  inject: [PrismaService],
};

@Module({
  providers: [dbosProviders],
  exports: [dbosProviders],
})
export class AppModule {}
```